runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/util/Util.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.http.endpoints.util
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+import aws.smithy.kotlin.runtime.util.net.isIpv4
+import aws.smithy.kotlin.runtime.util.net.isIpv6
+import aws.smithy.kotlin.runtime.util.net.isValidHostname
+import aws.smithy.kotlin.runtime.util.text.ensureSuffix
+import aws.smithy.kotlin.runtime.util.text.urlEncodeComponent
+
+@InternalApi
+public fun isSet(value: Any?): Boolean = value != null
+
+@InternalApi
+public fun not(value: Boolean): Boolean = !value
+
+@InternalApi
+public fun substring(value: String, start: Int, stop: Int, reverse: Boolean): String? =
+    when {
+        start >= stop || stop > value.length -> null
+        reverse -> value.substring(value.length - stop until value.length - start)
+        else -> value.substring(start until stop)
+    }
+
+@InternalApi
+public fun stringEquals(i: String, j: String): Boolean = i == j
+
+@InternalApi
+public fun isValidHostLabel(value: String, allowSubdomains: Boolean): Boolean =
+    if (allowSubdomains) value.isValidHostname() else value.split('.').all(String::isValidHostname)
+
+@InternalApi
+public fun uriEncode(value: String): String = value.urlEncodeComponent(formUrlEncode = false)
+
+@InternalApi
+public fun booleanEquals(i: Boolean, j: Boolean): Boolean = i == j
+
+@InternalApi
+public fun parseUrl(value: String): Url? {
+    val sdkUrl: aws.smithy.kotlin.runtime.http.Url
+    try {
+        sdkUrl = aws.smithy.kotlin.runtime.http.Url.parse(value)
+    } catch (e: Exception) {
+        return null
+    }
+
+    val isIpv6 = sdkUrl.host.isIpv6()
+    val authority = buildString {
+        append(if (isIpv6) "[${sdkUrl.host}]" else sdkUrl.host)
+        if (sdkUrl.port != sdkUrl.scheme.defaultPort) {
+            append(":${sdkUrl.port}")
+        }
+    }
+
+    return Url(
+        scheme = sdkUrl.scheme.protocolName,
+        authority,
+        path = sdkUrl.path,
+        normalizedPath = sdkUrl.path.ensureSuffix("/"),
+        isIp = isIpv6 || sdkUrl.host.isIpv4(),
+    )
+}
+
+@InternalApi
+public data class Url(
+    public val scheme: String,
+    public val authority: String,
+    public val path: String,
+    public val normalizedPath: String,
+    public val isIp: Boolean,
+)
```

Comment:
I think it's semantically clearer to have a discrete type here that maps from our SDK's public URL type. That said, extension `val`s would also be great, and probably more idiomatic.

The only downside I can see from that (correct me if I'm wrong) would be that we'd technically start recomputing the extension vals whenever they were evaluated in rulesets.

eg. `isIp` is checked once in the type mapping route. As an extension, which I imagine we'd implement as `public val ... get() = ...` would then perform the check per-access.

I definitely don't want to add these properties into the main type, that would mean we'd start duplicating some information across fields.

Reply:
How often would we be accessing these fields? Does precomputing them save us a significant amount of cycles?

Reply:
Per-request, since endpoints are now to be resolved in that way.

But, since I have ADT-ified the IP address per another suggestion above, the "repeated" check would now just be an `is` branch, instead of the address parsing we had before, so I'm more inclined to make them extensions at this point.

Reply:
I ended up not going the extension route - I noticed there would have been a conflict (the SEP specifies a scheme property, and so do we). Thought about it some more and it seems to make the most sense to keep these values discrete, considering their narrow use and similarity to our own SDK type.

Reply:
I'm not seeing a compelling reason to have a separate `Url` type here (and certainly not one named the same thing)